parser/src/dataset.rs

```diff
@@ -243,6 +286,43 @@ where
     }
 }
 
+
+impl<'s, S: 's, P, D> DataSetReader<S, P, D>
+where
+    P: Parse<dyn Read + 's>,
+    S: Read,
+{
+    fn update_seq_delimiters(&mut self) -> Option<DataToken> {
+        if let Some(sd) = self.seq_delimiters.last() {
+            if let Some(len) = sd.len.get() {
+                let eos = sd.base_offset + len as u64;
+                let bytes_read = self.parser.bytes_read();
+                if eos == bytes_read {
+                    // end of delimiter, as indicated by the element's length
+                    let token;
+                    match sd.typ {
+                        SeqTokenType::Sequence => {
+                            self.in_sequence = false;
+                            token = DataToken::SequenceEnd;
+                        },
+                        SeqTokenType::Item =>  {
+                            self.in_sequence = true;
+                            token = DataToken::ItemEnd;
+                        },
+                    }
+
+                    self.seq_delimiters.pop();
+                    return Some(token);
+                } else if eos < bytes_read {
+                    panic!("already read {} bytes, but end of sequence is @ {} bytes", bytes_read, eos);
```

Comment:
OK, I'm starting to think that this panic might not be a good idea. `eos` is influenced by the input DICOM data, so one could deliberately construct a DICOM file to panic a program. This kind of breach is very often worth fixing when known.

Keeping this note for myself to address this matter before merging. It should be as simple as turning it into a recoverable error.